File: backend/src/main/java/com/poeflipfinder/backend/framework/league/GggLeagueGateway.java
package com.poeflipfinder.backend.framework.league;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.poeflipfinder.backend.entity.League;
import com.poeflipfinder.backend.gateway.LeagueGateway;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import org.springframework.web.client.RestClient;

/**
 * The Frameworks & Drivers implementation of LeagueGateway, calling GGG's
 * public Leagues API (docs/DATA_SOURCES.md § League List). This is the
 * Anti-Corruption Layer boundary: everything above this class only ever
 * sees League entities, never this class's Raw* shapes.
 */
public class GggLeagueGateway implements LeagueGateway {

    private final RestClient restClient;
    private final ObjectMapper objectMapper;

    public GggLeagueGateway(RestClient.Builder restClientBuilder) {
        this.restClient = restClientBuilder.baseUrl("https://api.pathofexile.com").build();
        this.objectMapper = new ObjectMapper();
    }

    @Override
    public List<League> fetchLeagues() {
        String json = restClient.get().uri("/leagues").retrieve().body(String.class);
        return normalize(json);
    }

    /** Package-visible so the contract test can exercise parsing without a live HTTP call. */
    List<League> normalize(String json) {
        RawLeague[] rawLeagues = deserialize(json);
        return Arrays.stream(rawLeagues)
                .filter(raw -> !isSoloSelfFound(raw))
                .map(this::toEntity)
                .toList();
    }

    private RawLeague[] deserialize(String json) {
        try {
            return objectMapper.readValue(json, RawLeague[].class);
        } catch (IOException e) {
            throw new LeagueGatewayException("Failed to parse Leagues API response", e);
        }
    }

    private boolean isSoloSelfFound(RawLeague raw) {
        return hasRule(raw, "NoParties");
    }

    private League toEntity(RawLeague raw) {
        // hasExchangeActivity is not yet cross-checked against ingested market data
        // (docs/ARCHITECTURE.md § League Resolution step 3) -- wired in once the
        // Currency Exchange ingestion service exists.
        return new League(null, raw.id(), raw.name(), isMainlineCurrentChallengeLeague(raw), false);
    }

    // GGG flags every ruleset variant of the active challenge (softcore,
    // Hardcore, Ruthless, ...) with the same category.current=true, so that
    // flag alone matches more than one league -- only the plain variant with
    // no modifying rules is "the" mainline default from docs/PRD.md § 7.4.
    private boolean isMainlineCurrentChallengeLeague(RawLeague raw) {
        boolean isCurrentCategory = raw.category() != null && Boolean.TRUE.equals(raw.category().current());
        return isCurrentCategory && !hasRule(raw, "Hardcore") && !hasRule(raw, "HardMode");
    }

    private boolean hasRule(RawLeague raw, String ruleId) {
        return raw.rules() != null && raw.rules().stream().anyMatch(rule -> ruleId.equals(rule.id()));
    }

    // ignoreUnknown so fields GGG adds later (or ones we simply don't need,
    // e.g. realm/url/startAt) never break parsing -- only a field we
    // actually read being renamed or removed should ever fail this adapter.
    @JsonIgnoreProperties(ignoreUnknown = true)
    private record RawLeague(String id, String name, RawCategory category, List<RawRule> rules) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private record RawCategory(String id, Boolean current) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private record RawRule(String id, String name, String description) {
    }
}

File: app/lib/poe_flip_finder/gateways/ggg_league_gateway.ex
defmodule PoeFlipFinder.Gateways.GggLeagueGateway do
  @moduledoc """
  The imperative-shell implementation of `PoeFlipFinder.LeagueGateway`,
  calling GGG's public Leagues API (docs/DATA_SOURCES.md § League List).
  This is the Anti-Corruption Layer boundary: everything above this module
  only ever sees `PoeFlipFinder.League`, never GGG's raw JSON shape.
  """

  @behaviour PoeFlipFinder.LeagueGateway

  alias PoeFlipFinder.League

  @default_base_url "https://api.pathofexile.com"

  @impl true
  def fetch_leagues do
    url = base_url() <> "/leagues"

    case Req.get(url, retry: false) do
      {:ok, %Req.Response{status: 200, body: body}} -> {:ok, normalize(decode(body))}
      {:ok, %Req.Response{status: status}} -> {:error, {:unexpected_status, status}}
      {:error, reason} -> {:error, reason}
    end
  end

  defp base_url do
    Application.get_env(:poe_flip_finder, __MODULE__, [])[:base_url] || @default_base_url
  end

  defp decode(body) when is_binary(body), do: Jason.decode!(body)
  defp decode(body) when is_list(body), do: body

  @doc "Exposed so contract tests can exercise parsing without a live HTTP call."
  @spec normalize([map()]) :: [League.t()]
  def normalize(raw_leagues) do
    raw_leagues
    |> Enum.reject(&solo_self_found?/1)
    |> Enum.map(&to_entity/1)
  end

  defp solo_self_found?(raw), do: has_rule?(raw, "NoParties")

  defp to_entity(raw) do
    %League{
      id: nil,
      external_id: raw["id"],
      display_name: raw["name"],
      is_current: mainline_current_challenge_league?(raw),
      # Not yet cross-checked against ingested market data
      # (docs/ARCHITECTURE.md § League Resolution step 3) -- wired in once
      # the ingestion service exists (Phase 3/4 of the migration).
      has_exchange_activity: false
    }
  end

  # GGG flags every ruleset variant of the active challenge (softcore,
  # Hardcore, Ruthless, ...) with the same category.current=true, so that
  # flag alone matches more than one league -- only the plain variant with
  # no modifying rules is "the" mainline default from docs/PRD.md § 7.4.
  defp mainline_current_challenge_league?(raw) do
    is_current_category = get_in(raw, ["category", "current"]) == true
    is_current_category and not has_rule?(raw, "Hardcore") and not has_rule?(raw, "HardMode")
  end

  defp has_rule?(raw, rule_id) do
    Enum.any?(raw["rules"] || [], &(&1["id"] == rule_id))
  end
end
